src/alluvium/avro.py

```diff
@@ -0,0 +1,251 @@
+"""
+Defines the schema you expect from your CDC events.
+
+Provides synctactic sugar for Any types and Either Types.
+"""
+
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Any, List, Optional, Union
+
+
+class InvalidSchemaException(Exception):
+    def __init__(self, field_type, message, name=None):
+        self.error_field_type = field_type
+        self.name = name
+        super(InvalidSchemaException, self).__init__(
+            self, "Error with field <{} {}>. {}".format(field_type, name, message)
+        )
+
+
+class AvroPrimitiveType(Enum):
+    NULL = "null"
+    BOOLEAN = "boolean"
+    INTEGER = "int"
+    LONG = "long"
+    FLOAT = "float"
+    DOUBLE = "double"
+    BYTES = "bytes"
+    STRING = "string"
+
+
+class AvroComplexType(Enum):
+    RECORD = "record"
+    ENUM = "enum"
+    ARRAY = "array"
+    MAP = "map"
+    UNIONS = "unions"
+    FIXED = "fixed"
+
+
+class AvroCustomType(Enum):
+    ANY = "any"
+
+
+def resolve_primitive_type(avro_type: Union[AvroPrimitiveType, AvroCustomType]):
+    if isinstance(avro_type, AvroPrimitiveType):
+        return avro_type.value
+    if avro_type == AvroCustomType.ANY:
+        return AvroUnionSchema([primitive for primitive in AvroPrimitiveType]).avro_schema()
+    raise ValueError("Invalid base type provided")
+
+
+class FieldOrdering(Enum):
+    ASCENDING = "ascending"
+    DESCENDING = "descending"
+    IGNORE = "ignore"
+
+
+class AvroSchema(ABC):
+    def __init__(self, avro_type: Union[AvroPrimitiveType, AvroComplexType]):
+        self.type = avro_type
+
+    @abstractmethod
+    def avro_schema(self):
+        pass
+
+
+class AvroRecordField(object):
+    def __init__(
+        self,
+        name: str,
+        avro_type: Union[AvroSchema, AvroPrimitiveType],
+        doc: str = None,
+        default: Any = None,
+        order: FieldOrdering = None,
+        aliases: List[str] = None,
+    ):
+        self.name = name
+        self.avro_type = avro_type
+        self.doc = doc
+        self.default = default
+        self.order = order
+        self.aliases = aliases
+
+    def generate_avro_field(self):
+        base = {
+            "type": resolve_primitive_type(self.avro_type)
+            if isinstance(self.avro_type, (AvroPrimitiveType, AvroCustomType))
```

Comment:
If self.avro_type = AvroComplexType, then this might throw an error, since self.avro_type.avro_schema() is not defined.

Reply:
resolve_primitive_type yields either a primitive type's value or a schema which is a union of all primitives. So, related to the previous comment, is there some code design decision you have made for not including AvroComplexType as a record field type? What does AvroComplexType serve as in the schema?

Reply:
So `self.avro_type` is either a `AvroPrimitiveType` or a `AvroSchema`. You wouldn't be able to set a avro_type of complex type. That being said, this illuminates that I haven't captured the semantics correctly. 

a schema's `avro_type` is a string. A fields `avro_type` is either a primitive type OR another schema (because you have nested dicts and other wild stuff). I should probably name this something else. Maybe avro_field_type?

Reply:
Yes naming it avro_field_type would make it clearer to read. I understood the scheme after looking at the testing code.

Reply:
Yes `avro_field_type` might make things clearer in the RecordField class.

Reply:
👍